rust/codelist-rs/src/codelist.rs

```diff
@@ -190,6 +190,82 @@ impl CodeList {
     pub fn metadata(&self) -> &Metadata {
         &self.metadata
     }
+
+    /// Truncate codelist entries to 3 digits
+    ///
+    /// # Arguments
+    /// * `term_management` - How to handle ambiguous terms
+    ///
+    /// # Errors
+    /// * `CodeListError::CodeListNotTruncatable` - If the codelist is not ICD10
+    pub fn truncate_to_3_digits(
+        &mut self,
+        term_management: TermManagement,
+    ) -> Result<(), CodeListError> {
+        if !self.codelist_type.is_truncatable() {
+            return Err(CodeListError::CodeListNotTruncatable);
+        }
+
+        // Keep track of all the three-digit codes
+        let mut threes = self
+            .entries
+            .iter()
+            .filter(|entry| entry.code.len() == 3)
+            .map(|entry| entry.code.clone())
+            .collect::<HashSet<String>>();
+
+        let mut adds = vec![];
+        let mut removes = vec![];
+
+        for entry in self.entries.iter() {
+            // Codes of 3 or less will not be truncated
+            if entry.code.len() <= 3 {
+                continue;
+            }
+
+            // We'll remove this one later
+            removes.push(entry.clone());
+
+            // truncate the entry's code
+            let truncated_code = entry.code[..3].to_string();
+
+            // If we already have this one, then go on to the next one
+            if threes.contains(&truncated_code) {
```

Comment:
This is a really good edition. I initially thought that we don't need this because of the enties being `Hashset` but of course it matches across the whole entry, rather than just code, Nice catch.

Reply:
Yeah, I'm wondering if we will end up with entries morphing from a HashSet of CodeEntries to a HashMap with codes as the keys and the rest of the CodeEntry as values. Especially if both term and comment become optional as in #49.

Reply:
I reckon you're right. And, yes, a BTreeMap might be even better. I don't have a good feel for how big a codelist it would take before we cared about the performance difference between a HashMap and a BTreeMap, but if we care about sorting often enough, that would be the way to go. I like how Rust makes it easy to do that kind of refactoring, so there's another reason not to decide now.

Reply:
I created issue #52 to capture this idea. I don't think I have the permissions to do the merge myself.

Reply:
nice one, thanks. Will merge now.